protoc-gen-swagger/genswagger/template.go

```diff
@@ -1187,3 +1230,70 @@ func extractSwaggerOptionFromFileDescriptor(file *pbdescriptor.FileDescriptorPro
 	}
 	return opts, nil
 }
+
+func swaggerSchemaFromProtoSchema(s *swagger_options.Schema, reg *descriptor.Registry) swaggerSchemaObject {
+	ret := swaggerSchemaObject{
+		schemaCore:   schemaCore{},
+		ExternalDocs: protoExternalDocumentationToSwaggerExternalDocumentation(s.GetExternalDocs()),
+		// TODO(johanbrandhorst): Add more fields?
+	}
+	if s.GetJsonSchema().GetRef() != "" {
+		swaggerName := fullyQualifiedNameToSwaggerName(s.GetJsonSchema().GetRef(), reg)
+		if swaggerName != "" {
+			ret.schemaCore.Ref = "#/definitions/" + swaggerName
+		} else {
+			ret.schemaCore.Ref += s.GetJsonSchema().GetRef()
+		}
+	} else {
+		f, t := protoJSONSchemaTypeToFormat(s.GetJsonSchema().GetType())
+		ret.schemaCore.Format = f
+		ret.schemaCore.Type = t
+	}
+	return ret
+}
+
+func protoJSONSchemaTypeToFormat(in []swagger_options.JSONSchema_JSONSchemaSimpleTypes) (string, string) {
+	if len(in) == 0 {
+		return "", ""
+	}
+
+	// Can't support more than 1 type, just return the first element.
+	// This is due to an inconsistency in the design of the openapiv2 proto
+	// and that used in schemaCore. schemaCore uses the v3 definition of types,
+	// which only allows a single string, while the openapiv2 proto uses the OpenAPI v2
+	// definition, which defers to the JSON schema definition, which allows a string or an array.
+	// Sources:
+	// https://swagger.io/specification/#itemsObject
+	// https://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-5.5.2
+	switch in[0] {
+	case swagger_options.JSONSchema_UNKNOWN, swagger_options.JSONSchema_NULL:
+		return "", ""
+	case swagger_options.JSONSchema_OBJECT:
+		return "object", ""
+	case swagger_options.JSONSchema_ARRAY:
+		return "array", ""
+	case swagger_options.JSONSchema_BOOLEAN:
+		return "boolean", "boolean"
+	case swagger_options.JSONSchema_INTEGER:
+		return "integer", "int32"
+	case swagger_options.JSONSchema_NUMBER:
+		return "number", "double"
+	case swagger_options.JSONSchema_STRING:
+		// NOTE: in swagger specifition, format should be empty on string type
+		return "string", ""
+	default:
+		// Maybe panic?
+		return "", ""
+	}
+}
+
+func protoExternalDocumentationToSwaggerExternalDocumentation(in *swagger_options.ExternalDocumentation) *swaggerExternalDocumentationObject {
```

Comment:
Should https://github.com/johanbrandhorst/grpc-gateway/blob/97aefd3e41117e62982ef489dfb516a228a39ea3/protoc-gen-swagger/genswagger/template.go#L233-L241 and other places with `ExternalDocs` be combined with this?

Reply:
I think I squashed these.